mmv1/third_party/tgc_next/pkg/cai2hcl/converters/utils/utils.go

```diff
@@ -28,26 +29,60 @@ func ParseFieldValue(url string, name string) string {
 template: //bigquery.googleapis.com/projects/{{project}}/datasets/{{dataset_id}}
 assetName: //bigquery.googleapis.com/projects/my-project/datasets/my-dataset
 hclData: [project:my-project dataset_id:my-dataset]
+
+It also handles mutliple-path fragements.
+template: {{cluster}}/instances/{{instance_id}}
+assetName: //alloydb.googleapis.com/projects/ci-test-project/locations/us-central1/clusters/tf-test-cluster/instances/tf-test-instance
+{{cluster}} field covers projects/ci-test-project/locations/us-central1/clusters/tf-test-cluster
 */
 func ParseUrlParamValuesFromAssetName(assetName, template string, outputFields map[string]struct{}, hclData map[string]any) {
-	fragments := strings.Split(template, "/")
-	if len(fragments) < 2 {
-		// We need a field and a prefix.
-		return
-	}
-	fields := make(map[string]string) // keys are prefixes in URI, values are names of fields
-	for ix, item := range fragments[1:] {
-		if trimmed, ok := strings.CutPrefix(item, "{{"); ok {
-			if trimmed, ok = strings.CutSuffix(trimmed, "}}"); ok {
-				fields[fragments[ix]] = trimmed // ix is relative to the subslice
+	templateFragments := strings.Split(template, "/")
+	assetFragments := strings.Split(assetName, "/")
+
+	// Iterate through the fragments and match fields.
+	assetIx := 0
+	endAssetIx := 0
+	for templateIx := 0; templateIx < len(templateFragments); templateIx++ {
+		templateFragment := templateFragments[templateIx]
+
+		// Check if the template fragment is a field (e.g., {{project}})
+		if fieldName, isField := strings.CutPrefix(templateFragment, "{{"); isField {
+			fieldName, _ = strings.CutSuffix(fieldName, "}}")
+
+			// Find the end of this field in the template. The end is the next non-field fragment.
+			endTemplateIx := templateIx + 1
+			for endTemplateIx < len(templateFragments) && strings.HasPrefix(templateFragments[endTemplateIx], "{{") {
+				endTemplateIx++
 			}
-		}
-	}
-	fragments = strings.Split(assetName, "/")
-	for ix, item := range fragments[:len(fragments)-1] {
-		if fieldName, ok := fields[item]; ok {
+
+			if endTemplateIx != len(templateFragments) {
+				nextNonFieldFragment := templateFragments[endTemplateIx]
+				for {
+					if endAssetIx == len(assetFragments) || assetFragments[endAssetIx] == nextNonFieldFragment {
+						break
+					} else {
+						endAssetIx++
+					}
+				}
+			} else {
+				endAssetIx = len(assetFragments)
+			}
```

Comment:
I'm finding this block of code a little hard to follow given that it has multiple loops nested inside a conditional inside a loop. Would it be possible to move some of the logic into a helper function?

Reply:
Sure. Added a helper function `getEndAssetIx`.

Feel free to let me know if more helper functions are needed.